File: Astromaximum/Classes/AmaxLocationListController.m
//
//  AmaxLocationListController.m
//  
//
//  Created by admin on 03.12.12.
//  Copyright (c) 2012 __MyCompanyName__. All rights reserved.
//

#import "AmaxLocationListController.h"
#import "Astromaximum-Swift.h"
#import "AmaxPrefs.h"

@implementation AmaxLocationListController
@synthesize tableView;

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(didSelectLocation:)];
        self.title = NSLocalizedString(@"Current location", @"Current location");
        self->mDataProvider = [AmaxDataProvider sharedInstance];
    }
    return self;
}

- (void)didReceiveMemoryWarning
{
    // Releases the view if it doesn't have a superview.
    [super didReceiveMemoryWarning];
    
    // Release any cached data, images, etc that aren't in use.
}

#pragma mark - View lifecycle

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    workingLocationIndex = 0;
    for (long i = 0; i < [mDataProvider.mSortedLocationKeys count]; ++i) {
        if ([mDataProvider.mLocationId isEqualToString:[mDataProvider.mSortedLocationKeys objectAtIndex:i]]) {
            workingLocationIndex = i;
            break;
        }
    }
    NSIndexPath* path = [NSIndexPath indexPathForRow:workingLocationIndex inSection:0];
    [self tableView:self.tableView didSelectRowAtIndexPath:path];
    [self.tableView scrollToRowAtIndexPath:path atScrollPosition:UITableViewScrollPositionMiddle animated:true];
    self.navigationItem.rightBarButtonItem.enabled = false;
}

- (void)viewDidDisappear:(BOOL)animated
{
    NSIndexPath* path = [NSIndexPath indexPathForRow:workingLocationIndex inSection:0];
    [self tableView:self.tableView didSelectRowAtIndexPath:path];
    selectedLocationIndex = -1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return mDataProvider.mSortedLocationKeys.count;
}

// Customize the appearance of table view cells.
- (UITableViewCell *)tableView:(UITableView *)tableView1 cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"Cell";
    
    UITableViewCell *cell = [tableView1 dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:CellIdentifier];
    }
    
    cell.textLabel.text = [mDataProvider.mLocations objectForKey:[mDataProvider.mSortedLocationKeys objectAtIndex:indexPath.row]];
    
    NSString *key = [mDataProvider.mSortedLocationKeys objectAtIndex:indexPath.row];
    cell.detailTextLabel.text = key;

    cell.accessoryType = selectedLocationIndex == indexPath.row ? UITableViewCellAccessoryCheckmark : UITableViewCellAccessoryNone;

    return cell;
}

- (void)tableView:(UITableView *)tableView1 didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView1 deselectRowAtIndexPath:indexPath animated:NO];
    
    if (selectedLocationIndex == indexPath.row)
        return;
    
    UITableViewCell *newCell = [tableView1 cellForRowAtIndexPath:indexPath];
    newCell.accessoryType = UITableViewCellAccessoryCheckmark;
    
    NSIndexPath *oldIndexPath = [NSIndexPath indexPathForRow:selectedLocationIndex inSection:0];
    UITableViewCell *oldCell = [tableView1 cellForRowAtIndexPath:oldIndexPath];
    oldCell.accessoryType = UITableViewCellAccessoryNone;

    //NSLog(@"didSelect new %d, old %d, selected %d, working %d", [indexPath indexAtPosition:1], [oldIndexPath indexAtPosition:1], selectedLocationIndex, workingLocationIndex);
    
    selectedLocationIndex = indexPath.row;
    self.navigationItem.rightBarButtonItem.enabled = selectedLocationIndex != workingLocationIndex;
}

- (IBAction)didSelectLocation:(id)sender
{
    workingLocationIndex = selectedLocationIndex;
    [mDataProvider loadLocationByIdWithLocationId: [mDataProvider.mSortedLocationKeys objectAtIndex:workingLocationIndex]];
    [self.navigationController popViewControllerAnimated:YES];
}

@end

File: Astromaximum/Classes/AmaxLocationListController.swift
//
//  AmaxLocationListController.m
//  
//
//  Created by admin on 03.12.12.
//  Copyright (c) 2012 __MyCompanyName__. All rights reserved.
//

import UIKit

class AmaxLocationListController : UIViewController, UITableViewDataSource, UITableViewDelegate {

    private var mDataProvider:AmaxDataProvider!
    private var workingLocationIndex = -1
    private var selectedLocationIndex = -1
    @IBOutlet weak var tableView1: UITableView!

    override init(nibName: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName:nibName, bundle:nibBundleOrNil)
        navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(self.didSelectLocation(sender:)))
        title = NSLocalizedString("pref_current_location", comment: "Current location")
        mDataProvider = AmaxDataProvider.sharedInstance
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func didReceiveMemoryWarning() {
        // Releases the view if it doesn't have a superview.
        super.didReceiveMemoryWarning()

        // Release any cached data, images, etc that aren't in use.
    }

    // MARK: - View lifecycle

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        workingLocationIndex = 0
        for i in 0 ..< mDataProvider.mSortedLocationKeys.count {
            if (mDataProvider.mLocationId == mDataProvider.mSortedLocationKeys[i].0) {
                workingLocationIndex = i
                break
            }
         }
        let path = IndexPath(row: workingLocationIndex, section: 0)
        tableView(tableView1, didSelectRowAt: path)
        tableView1.scrollToRow(at: path, at: .middle, animated: true)
        navigationItem.rightBarButtonItem?.isEnabled = false
    }

    override func viewDidDisappear(_ animated:Bool) {
        let path = IndexPath(row: workingLocationIndex, section: 0)
        tableView(tableView1, didSelectRowAt: path)
        selectedLocationIndex = -1
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return mDataProvider.mSortedLocationKeys.count
    }

    // Customize the appearance of table view cells.
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let CellIdentifier = "Cell"

        var cell = tableView1.dequeueReusableCell(withIdentifier: CellIdentifier)
        if cell == nil {
            cell = UITableViewCell(style: .subtitle, reuseIdentifier: CellIdentifier)
        }

        let loc = mDataProvider.mLocations[mDataProvider.mSortedLocationKeys[indexPath.row].0]!

        cell?.textLabel?.text = loc.mCity
        cell?.detailTextLabel?.text = loc.mCountry
        if !loc.mState.isEmpty {
            cell?.detailTextLabel?.text! += ", " + loc.mState
        }

        cell?.accessoryType = selectedLocationIndex == indexPath.row ? .checkmark : .none

        return cell!
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView1.deselectRow(at: indexPath, animated: false)

        if selectedLocationIndex == indexPath.row {
            return
        }

        let newCell = tableView1.cellForRow(at: indexPath)
        newCell?.accessoryType = .checkmark

        let oldIndexPath = IndexPath(row: selectedLocationIndex, section:0)
        if let oldCell = tableView1.cellForRow(at: oldIndexPath) {
            oldCell.accessoryType = .none
        }
        //NSLog(@"didSelect new %d, old %d, selected %d, working %d", [indexPath indexAtPosition:1], [oldIndexPath indexAtPosition:1], selectedLocationIndex, workingLocationIndex);

        selectedLocationIndex = indexPath.row
        self.navigationItem.rightBarButtonItem?.isEnabled = selectedLocationIndex != workingLocationIndex
    }

    @IBAction func didSelectLocation(sender:AnyObject!) {
        workingLocationIndex = selectedLocationIndex
        mDataProvider.loadLocationById(locationId:  mDataProvider.mSortedLocationKeys[workingLocationIndex].0)
        navigationController?.popViewController(animated: true)
    }
}
